iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/IotHubConnectionString.java

```diff
@@ -20,17 +20,29 @@
 {
     /** The hostName attribute name in a connection string. */
     private static final String HOSTNAME_ATTRIBUTE = "HostName=";
+
     /** The device ID attribute name in a connection string. */
     private static final String DEVICE_ID_ATTRIBUTE = "DeviceId=";
+
     /** The shared access key attribute name in a connection string. */
     private static final String SHARED_ACCESS_KEY_ATTRIBUTE = "SharedAccessKey=";
+
     /** The shared access signature attribute name in a connection string. */
     private static final String SHARED_ACCESS_TOKEN_ATTRIBUTE = "SharedAccessSignature=";
 
+    /** The module ID attribute name in a connection string. */
     private static final String MODULE_ID_ATTRIBUTE = "ModuleId=";
 
+    /**
+     * IP address or internet name of the host machine working as a device or protocol gateway.
+     * Used when communicating with Azure Edge devices.
+     * */
     private static final String GATEWAY_HOST_NAME_ATTRIBUTE = "GatewayHostName=";
 
+    /** Used for E4K. */
+    private static final String MQTT_GATEWAY_HOST_NAME_ATTRIBUTE = "MqttGatewayHostName=";
```

Comment:
The .NET SDK has some client-side regex validations for the values parsed from the connection string. I see that the java SDK doesn't do any such validations so it isn't an issue here, but we might want to get some information on the character constraints for this hostname.

This is our existing regex validation for both IoT hub hostname and gateway hostname (edge): "[a-zA-Z0-9_\-\.]+$"

Reply:
Yeah, this sounds fair to add.

Reply:
I'm not sure I understand why we would want to validate a host name like this. If the host name is incorrect, then the user will see a HostNotFoundException or something similar.

Reply:
> This is our existing regex validation for both IoT hub hostname and gateway hostname (edge): "[a-zA-Z0-9_-.]+$"
Do we define these character constraints ourselves or we refer to guidance somewhere else?

Reply:
> I'm not sure I understand why we would want to validate a host name like this. If the host name is incorrect, then the user will see a HostNotFoundException or something similar.

I agree with that and that's what we have in v2, but unfortunately this is the existing behavior in v1. We can choose to not add the validation for mqtt gateway hostname but that would change the "expected" behavior of how we treat the rest of the components of a connection string.

Reply:
Ah, I see your points.

Reply:
Sounds good then.